src/ucp/api/ucp.h

```diff
@@ -2457,6 +2517,77 @@ ucs_status_ptr_t ucp_tag_send_sync_nb(ucp_ep_h ep, const void *buffer, size_t co
                                       ucp_send_callback_t cb);
 
 
+/**
+ * @ingroup UCP_COMM
+ * @brief Non-blocking tagged-send operation
+ *
+ * This routine sends a messages that is described by the local address @a
+ * buffer, size @a count object to the destination endpoint @a ep. Each
+ * message is associated with a @a tag value that is used for message
+ * matching on the @ref ucp_tag_recv_nb "receiver". The routine is
+ * non-blocking and therefore returns immediately, however the actual send
+ * operation may be delayed. The send operation is considered completed when
+ * it is safe to reuse the source @e buffer. If the send operation is
+ * completed immediately the routine return UCS_OK and the call-back function
```

Comment:
@yosefe this description very much depends on ucp_op_attr_t ? Seems like it is copy-n-past previous document and does not represent new request options ? We also have to specify how request can be provided, etc.

Reply:
@shamisp the request options are documented in ucp_request_param_t:
`@param [in]  param       Operation parameters, see @ref ucp_request_param_t`

The request options documentation should be separate, and not for every API function, since it will be the same for all APIs (tag, put/get, stream, etc)

Reply:
@yosefe we cannot just copy and paint text from previous operation. The completion semantics changed and it is now based on optional argument passed through parameter.
What I'm saying is that it has to be reflected in function description.

Reply:
@shamisp the semantics and the completion semantics of the operation have not changed. Only the method of passing input parameters has changed, which is documented for the input parameter itself.
Would you expect that every UCP API function (tag_send_nbx, put_nbx, get_nbx, etc)  will have copy-paste of the same description of request parameters structure? Why is it better than just documenting that structure?

Reply:
@yosefe For example, @ ucp_request_param_t does not have completion semantics description. It is hidden under @ ucp_op_attr_t which is second level of re-direction from the function. And even in this case, the description of completion semantics is less than a sentence and not very descriptive. For each one of the configuration we have to describe what error codes are returned and what is the meaning. I also this we should introduce code examples, similar to https://github.com/openucx/ucx/blob/master/src/ucp/api/ucp.h#L1874
You shared similar examples on the mail threads , so we can reuse those. Hopefully it clarifies.

Reply:
@shamisp so to clarify, we need to extend the documentation of @ ucp_request_param_t to clarify how request completion is reported for different combination of parameters?